docs/proposals/20200511-clusterctl-extensible-template-processing.md

```diff
@@ -0,0 +1,297 @@
+within clusterctl so that it can be used effectively with other templating
+tools and continue to provide a seamless user experience.
+
+## Motivation
+
+To support a Cluster API that is cloud agnostic, operators need to deploy
+workloads in a multi-cloud environment. Operators and providers may need to
+create and _manage_ multiple cluster templates across a matrix of options and
+configurations. To manage the explosion of yaml manifests, operators may want
+to explore templating tools to minimize duplication and assert type
+correctness.
+
+Users looking to deploy workload clusters may be overwhelmed by the number of
+variables needed to configure a workload cluster. Some of these options may be
+circumvented via defaults provided by the operators as part of the templating
+tool.
+
+This proposal aims to extend clusterctl’s current templating mechanism of
+simple variable substitution to support other templating formats natively,
+thus benefiting the operators to deploy workload clusters with minimal
+configuration.
+
+### Goals
+
+- Clusterctl will not be changing the way it gets values for yaml variables.
+  That is, it will get values from environment variables and the config file.
+- To provide an interface such that other templating solutions may be
+  supported.
+
+### Non-Goals
+
+- To disrupt upstream templating solutions such as kustomize.
+- To replace Clusterctl default mode of processing templates, aka
+  `SimpleYamlProcessor`.
+
+### Future Work
+
+- To automatically detect template files for the appropriate template engine.
+- To use the configuration file to determine which provider should use
+  prefered templating mechanisms.
+
+## Proposal
+
+This section outlines the high level work that needs to be done in order to
+accomplish the task of supporting other templating tools.
+
+### User Stories
+
+- As a clusterctl user, I would like to preserve the current behavior of using
+  existing workload cluster templates with simple variable substitution from
+  the clusterctl config and environment variables.
+- As an infrastructure provider, I would like to provide templated yaml with
+  defaults and have that be consumed by `clusterctl config cluster` to
+  generate cluster manifests.
+- As a developer, I would like to easily provide a way to extend clusterctl
+  with my preferred templating solution.
+
+### Implementation Details/Notes/Constraints
+
+Currently the clusterctl command fetches yaml for the provider components
+(yaml defining the management cluster components) and the workload cluster
+templates.
+
+#### UX for using Extensible Templating Tools
+For the first iteration, we are opting to use flags to activate the use of a
+specific templating tool as follows,
+
+```
+# Retrieve variables required by the preferred template processor
+clusterctl config cluster <cluster-name> --infrastructure <provider-name> --list-variables --experimental-template-engine=templateProcessor
```

Comment:
As discussed in zoom, I don't think we need to add the `--experimental-template-engine=templateProcessor` option in the CLI in this phase since we won't be implementing any specific processors as part of this initial proposal.

Reply:
Thanks for the feedback @CecileRobertMichon! I've updated the CAEP to reflect that we are currently exposing library changes only so the existing behavior and UX won't change.